src/vec_bcd_ppc.h

```diff
@@ -554,9 +1400,478 @@ vec_bcdsub (vBCD_t a, vBCD_t b)
+  if (vec_all_eq((vui32_t) t, mz))
+    t = _BCD_CONST_ZERO;
+#endif
+  return (t);
+}
+
+/** \brief Decimal Subtract Extended Signed Modulo Quadword.
+ *
+ * Two Signed 31 digit values and a signed carry-in are subtracted
+ * (a - b- c) and lower 31 digits of the subtraction is returned.
+ * Overflow (carry-out) is ignored.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   |  26   | 1/cycle  |
+ * |power9   |   6   | 2/cycle  |
+ *
+ * @param a a 128-bit vector treated as a signed BCD 31 digit value.
+ * @param b a 128-bit vector treated as a signed BCD 31 digit value.
+ * @param c a 128-bit vector treated as a signed BCD carry with values
+ * -1, 0, or +1.
+ * @return a 128-bit vector which is the lower 31 digits of (a + b + c).
+ */
+static inline vBCD_t
+vec_bcdsubesqm (vBCD_t a, vBCD_t b, vBCD_t c)
+{
+  vBCD_t t;
+#ifdef _ARCH_PWR8
+  t = vec_bcdsub (vec_bcdsub (a, b), c);
+#else
+  const vui32_t mz = CONST_VINT128_W (0, 0, 0, 0x0000000d);
+  _Decimal128 d_t;
+  d_t = vec_BCD2DFP (a) - vec_BCD2DFP (b) - vec_BCD2DFP (c);
+  t = vec_DFP2BCD(d_t);
+  // fix up spurious negative zeros
+  if (vec_all_eq((vui32_t) t, mz))
+    t = _BCD_CONST_ZERO;
+#endif
+  return (t);
+}
+
+/** \brief Decimal Unsigned Shift.
+ * Shift a vector unsigned BCD value, left or right a variable
+ * amount of digits (nibbles).
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   | 12-14 | 1/cycle  |
+ * |power9   |   3   | 2/cycle  |
+ *
+ * @param vra 128-bit vector treated as a signed BCD 32 digit value.
+ * @param vrb Digit shift count in vector byte 7.
+ * @return a 128-bit vector BCD value shifted right digits.
+ */
+static inline vBCD_t
+vec_bcdus (vBCD_t vra, vi8_t vrb)
+{
+  vBCD_t vrt;
+#ifdef _ARCH_PWR9
+  __asm__(
+      "bcdus. %0,%1,%2;\n"
+      : "=v" (vrt)
+      : "v" (vrb),
+      "v" (vra)
+      : "cr6" );
+#else
+  const vi8_t zero = vec_splat_s8(0);
+  vi8_t shd = vec_splat (vrb, VEC_BYTE_L_DWH);
+  vui128_t t;
+  // Multiply digit shift by 4 to get bit shift count
+  shd = vec_add (shd, shd);
+  shd = vec_add (shd, shd);
+  t = (vui128_t) vra;
+  // Compare shift positive or negative
+  if (vec_all_ge(shd, zero))
+    {
+      // Positive, shift left
+      t = vec_slq (t, (vui128_t) shd);
+    }
+  else
+    {
+      // Negative, shift right by absolute value
+      shd = vec_sub (zero, shd);
+      t = vec_srq (t, (vui128_t) shd);
+    }
+  vrt = (vBCD_t) t;
+#endif
+  return (vrt);
+}
+
+/** \brief Combined Decimal Add & write Carry Signed Quadword.
+ *
+ * Two Signed 31 digit BCD values are added, and the carry-out
+ * (the high order 32nd digit) of the sum is generated.
+ * Botht the sum and the carry are returned.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   | 15-24 | 1/cycle  |
+ * |power9   | 6-15  | 2/cycle  |
+ *
+ * @param *cout a pointer to a 128-bit vector to recieve the BCD carry-out.
```

Comment:
nit: s/*cout/cout/
cout is the "pointer", not "*cout".
There are a couple more like this, below.

Reply:
OK